File: typescript/src/problems401_500/island-perimeter.ts
class Problem {
  islandPerimeter(grid: number[][]): number {
    throw new Error(`Not implemented (${grid}).`);
  }
}

class Default extends Problem {
  islandPerimeter(grid: number[][]): number {
    const rows: number = grid.length;
    const cols: number = grid[0].length;

    let islands: number = 0;
    let neighbors: number = 0;
    for (let row = 0; row < rows; row++) {
      for (let col = 0; col < cols; col++) {
        if (grid[row][col] !== 1) {
          continue;
        }
        islands++;
        if (row - 1 >= 0 && grid[row - 1][col] === 1) {
          neighbors++;
        }
        if (row + 1 < rows && grid[row + 1][col] === 1) {
          neighbors++;
        }
        if (col - 1 >= 0 && grid[row][col - 1] === 1) {
          neighbors++;
        }
        if (col + 1 < cols && grid[row][col + 1] === 1) {
          neighbors++;
        }
      }
    }
    return islands * 4 - neighbors;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems401_500/island_perimeter.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def island_perimeter(self, grid: list[list[int]]) -> int:
        pass


class Default(Problem):
    @override
    def island_perimeter(self, grid: list[list[int]]) -> int:
        rows: int = len(grid)
        cols: int = len(grid[0])

        islands: int = 0
        neighbors: int = 0
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] != 1:
                    continue
                islands += 1
                if row - 1 >= 0 and grid[row - 1][col] == 1:
                    neighbors += 1
                if row + 1 < rows and grid[row + 1][col] == 1:
                    neighbors += 1
                if col - 1 >= 0 and grid[row][col - 1] == 1:
                    neighbors += 1
                if col + 1 < cols and grid[row][col + 1] == 1:
                    neighbors += 1
        return islands * 4 - neighbors


SOLUTIONS: list[Problem] = [Default()]
